In [2]:
#import library to load the dataset, the dataset contains a date, so we parse the date as date.
#In this ML task, the date column was not used though.
import pandas as pd
df=pd.read_csv('energydata_complete.csv',parse_dates=['date'])
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0 2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1 2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2 2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3 2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4 2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

### Q1
From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius 
**(x = T2)** and the temperature outside the building **(y = T6)**.

**What is the Root Mean Squared error in 3 d.p?**

In [3]:
#this is a simple linear regression problem as there is just one predictor ('T2')
#first, import the libraries needed
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [4]:
#select the columns needed to solve the problem
X=df[['T2']]
y=df['T6']

#instantiate the model and fit to the dataset
model=LinearRegression()
model.fit(X,y)

#generate the predictions made by the model
predict=model.predict(X)

#calculate the root mean squared error (by finding the square root of the mean_squared_error)
print(f"Root Mean Squared Error = {round(np.sqrt(mean_squared_error(y,predict)),3)}")

Root Mean Squared Error = 3.644


### Q2
Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. 

Answer the following questions:

**What is the Mean Absolute Error (in three decimal places) for the  training set?**

In [5]:
#import the libraries to answer the question

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

#drop the columns not needed in answering the question
X=df.drop(columns=['date','lights','Appliances'])

y=df['Appliances']

In [6]:
#split the dataset into train and test data, use 30% of the data for testing.

X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=.3, random_state=42)

print(f"Training data has {len(X_train)} rows")
print(f"Testing data has {len(X_test)} rows")

Training data has 13814 rows
Testing data has 5921 rows


In [7]:
#Normalize the data using the MinMaxScaler to avoid some features from having more effect in prediction of the target

min_max=MinMaxScaler()
X_train_scaled=min_max.fit_transform(X_train)

X_test_scaled=min_max.transform(X_test)

In [8]:
#create a new linear regression model
lr_model=LinearRegression()
lr_model.fit(X_train_scaled, y_train)

LinearRegression()

In [9]:
#calculate the mean_absolute_error of the training dataset
#import the library for the solution
from sklearn.metrics import mean_absolute_error

pred=lr_model.predict(X_train_scaled)
print(f"Mean Absolute Error of training data = {round(mean_absolute_error(pred,y_train),3)}")

Mean Absolute Error of training data = 53.742


### Q4
**What is the Root Mean Squared Error (in three decimal places) for the training set?**

In [10]:
#we take the square root of the mean squared error

print(f"Root Mean Squared Error of training data = {round(np.sqrt(mean_squared_error(pred,y_train)),3)}")

Root Mean Squared Error of training data = 95.216


### Q5
**What is the Mean Absolute Error (in three decimal places) for test set?**

In [11]:
test_pred=lr_model.predict(X_test_scaled)

print(f"Mean Absolute Error of test data = {round(mean_absolute_error(test_pred,y_test),3)}")

Mean Absolute Error of test data = 53.643


### Q6
**What is the Root Mean Squared Error (in three decimal places) for test set?**

In [12]:
print(f"Root Mean Squared Error for test data = {round(np.sqrt(mean_squared_error(test_pred,y_test)),3)}")

Root Mean Squared Error for test data = 93.64


### Q7

**Did the Model above overfit to the training set**

Since the training root_mean_squared_error (95.216) is similar or close to the testing root_mean_squared_error (93.64),the model is performing equally on both the training and the test data.

Therefore, **the model is not Overfitting**

In [13]:
#alternatively, we can check the accuracy (r^2) of the model on both the training and test dataset
training_accuracy=lr_model.score(X_train_scaled,y_train)
testing_accuracy=lr_model.score(X_test_scaled,y_test)

print(
    f"Training accuracy ({round(training_accuracy,3)})", 
    f"is approximately equal to testing accuracy ({round(testing_accuracy,3)})",
    "The model is therefore not overfitting")

Training accuracy (0.145) is approximately equal to testing accuracy (0.149) The model is therefore not overfitting


### Q8

Train a ridge regression model with default parameters. 

**Is there any change to the root mean squared error (RMSE) when evaluated on the test set?**

In [14]:
#import the library needed to solve the problem
from sklearn.linear_model import Ridge

#calling the various parameters from previous solutions
X_train=X_train_scaled
y_train=y_train
X_test=X_test_scaled
y_test=y_test

ridge_model=Ridge()
ridge_model.fit(X_train,y_train)

pred_test=ridge_model.predict(X_test)
print("Root Mean Squared Error for Ridge model on test dataset = ",
      f"{round(np.sqrt(mean_squared_error(pred_test, y_test)),3)}")
print(f"Root Mean Squared Error for test data = {round(np.sqrt(mean_squared_error(test_pred,y_test)),3)}")

#They are aproximately the same and there is no significant change

Root Mean Squared Error for Ridge model on test dataset =  93.709
Root Mean Squared Error for test data = 93.64


### Q9

Train a lasso regression model with default value and obtain the new feature weights with it. 

**How many of the features have non-zero feature weights?**

In [15]:
#import the library needed
from sklearn.linear_model import Lasso

#calling the various parameters from previous solutions
#using the features before splitting into train and test
X=X
y=y
# Create and fit the Lasso regression model with default values
lasso_model = Lasso()
lasso_model.fit(X, y)

# Get the feature weights (coefficients)
feature_weights = lasso_model.coef_

# Count the number of features with non-zero feature weights
non_zero_features = sum(feature_weights != 0) 
#since the returned output is a boolean, the sum function will sum all the feature weights which are not zero.

print("Number of features with non-zero feature weights:", non_zero_features)


Number of features with non-zero feature weights: 20


### Q10

**What is the new RMSE with the Lasso Regression on the test set?**

In [16]:
X=X
y=y

#generating a new split dataset
X_train,X_test, y_train, y_test=train_test_split(X,y, test_size=.3, random_state=42)

# Assuming you have already split your data into X_train, X_test, y_train, and y_test
# Create and train the Lasso Regression model with default values
lasso_model = Lasso()
lasso_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = lasso_model.predict(X_test)

# Calculate the RMSE
rmse = np.sqrt(mean_squared_error(y_pred, y_test))

print("Root Mean Square Error (RMSE) with Lasso Regression on the test set:", round(rmse,3))

Root Mean Square Error (RMSE) with Lasso Regression on the test set: 93.892
